src/codegen/generators/statement_generator.rs

```diff
@@ -325,117 +328,102 @@ impl<'a, 'b> StatementCodeGenerator<'a, 'b> {
+
+        let step_ty = by_step.as_ref().map(|it| {
+            self.register_debug_location(it);
+            self.annotations.get_type_or_void(it, self.index)
+        });
+
+        let eval_step = || {
+            step_ty.map_or_else(
+                || self.llvm.create_const_numeric(&ll_ty, "1", SourceLocation::undefined()),
+                |step_ty| {
+                    let step = exp_gen.generate_expression(by_step.as_ref().unwrap())?;
+                    Ok(cast_if_needed!(exp_gen, ty, step_ty, step, None))
+                },
+            )
+        };
 
-        builder.build_conditional_branch(to_i1(or_eval.into_int_value(), builder), for_body, continue_block);
+        let predicate_incrementing = context.append_basic_block(current_function, "predicate_sle");
+        let predicate_decrementing = context.append_basic_block(current_function, "predicate_sge");
+        let loop_body = context.append_basic_block(current_function, "loop");
+        let increment = context.append_basic_block(current_function, "increment");
+        let afterloop = context.append_basic_block(current_function, "continue");
 
-        //Enter the for loop
-        builder.position_at_end(for_body);
-        let body_generator = StatementCodeGenerator {
-            current_loop_exit: Some(continue_block),
-            current_loop_continue: Some(increment_block),
+        self.generate_assignment_statement(counter, start)?;
+        let counter = exp_gen.generate_lvalue(counter)?;
+
+        // generate loop predicate selector. since `STEP` can be a reference, this needs to be a runtime eval
+        // XXX(mhasel): IR could possibly be improved by generating phi instructions.
+        //              Candidate for frontend optimization for builds without optimization when `STEP`
+        //              is a compile-time constant
+        let is_incrementing = builder.build_int_compare(
+            inkwell::IntPredicate::SGT,
+            eval_step()?.into_int_value(),
+            self.llvm.create_const_numeric(&ll_ty, "0", SourceLocation::undefined())?.into_int_value(),
+            "is_incrementing",
+        );
+        builder.build_conditional_branch(is_incrementing, predicate_incrementing, predicate_decrementing);
+        // generate predicates for incrementing and decrementing counters
+        let generate_predicate = |predicate| {
+            builder.position_at_end(match predicate {
+                inkwell::IntPredicate::SLE => predicate_incrementing,
+                inkwell::IntPredicate::SGE => predicate_decrementing,
+                _ => unreachable!(),
+            });
+
+            let end = exp_gen.generate_expression_value(end).unwrap();
+            let end_value = match end {
+                super::expression_generator::ExpressionValue::LValue(ptr) => builder.build_load(ptr, ""),
+                super::expression_generator::ExpressionValue::RValue(val) => val,
+            };
```

Comment:
nit: Can we remove the fully qualified paths, `ExpressionValue::LValue` or `LValue` instead of what RA auto pasted